web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java

```diff
@@ -129,14 +155,39 @@ private long getTokenExpiryTime(String[] cookieTokens) {
 		}
 	}
 
+	private RememberMeHashingAlgorithm getTokenHashingAlgorithm(String[] cookieTokens) {
+		if (cookieTokens.length == 3) {
+			return RememberMeHashingAlgorithm.UNSET;
+		}
+		return RememberMeHashingAlgorithm.findByIdentifier(cookieTokens[2]).orElseThrow(IllegalArgumentException::new);
+	}
+
 	/**
-	 * Calculates the digital signature to be put in the cookie. Default value is MD5
-	 * ("username:tokenExpiryTime:password:key")
+	 * Calculates the digital signature expected in the cookie if no algorithm is declared
+	 * or configured.
+	 *
+	 * Default value is MD5("username:tokenExpiryTime:password:key")
 	 */
 	protected String makeTokenSignature(long tokenExpiryTime, String username, String password) {
+		return makeTokenSignature(tokenExpiryTime, username, password, RememberMeHashingAlgorithm.MD5);
+	}
+
+	/**
+	 * Calculates the digital signature expected in the cookie when a hashing algorithm is
+	 * declared or configured.
+	 *
+	 * Default value is the hashing algorithm applied to
+	 * "username:tokenExpiryTime:password:key"
+	 * @since 5.5
+	 */
+	protected String makeTokenSignature(long tokenExpiryTime, String username, String password,
+			RememberMeHashingAlgorithm algorithm) {
+		if (algorithm == RememberMeHashingAlgorithm.UNSET) {
+			return makeTokenSignature(tokenExpiryTime, username, password);
+		}
```

Comment:
```suggestion
		if (algorithm == null) {
			algorithm = DEFAULT_HASHING_ALGORITHM;
		}
```

By doing this you can get rid of the `protected String makeTokenSignature(long tokenExpiryTime, String username, String password)` method.